ops/model.py

```diff
@@ -797,6 +927,267 @@ def __init__(self, name: str, address_info: '_AddressDict'):
         # TODO: expose a hostname/canonical name for the address here, see LP: #1864086.
 
 
+class _Secret:
+    """Secret object.
+
+    Not meant to be instantiated directly by charm code.
+    Typically you should obtain a Secret instance by creating one via your Unit or your Application,
+    or get one (typically one you don't own) from your Model.
+    """
+    secret_key_re = re.compile(r"^([a-z](?:-?[a-z0-9]){2,})$")
+
+    def __init__(self, backend: '_ModelBackend',
+                 secret_id: Optional[str] = None,
+                 label: Optional[str] = None,
+                 revision: Optional[int] = None, am_owner: bool = False):
+        if not label and not secret_id:
+            raise RuntimeError('Provide at least one of (`label`|`secret_id`).')
+        if am_owner and revision is None:
+            # This should only happen in testing, when one instantiates Secret manually.
+            logger.error('secret {!r} (owned by this unit) has no revision!'
+                         ' Things might malfunction'.format(secret_id))
+
+        self._backend = backend
+
+        if isinstance(secret_id, str):
+            if not secret_id.startswith('secret:'):  # todo: regex validation?
+                raise InvalidSecretIDError('secret secret ID {!r} should start with `secret:`'.format(secret_id))
+            if secret_id.strip() != secret_id:
+                logger.warning('invalid characters in secret ID '
+                               '{!r}: something wrong'.format(secret_id))  # FIXME
+                secret_id = secret_id.strip()
+
+        self._id = secret_id
+        self._label = label
+        self._revision = revision
+        self._am_owner = am_owner
+
+    @property
+    def id(self) -> str:
+        """Unique identifier for this secret."""
+        return self._id
+
+    @property
```

Comment:
It's standard in `ops` to make public attributes like these "read-only" (_a la python_). Else one might have the reasonable expectation that setting the label will also update it with juju, while it will not.
Since _Secret is a class users will interact with, it seemed sensible to guard against careless usage.